### PR TITLE
Upgrade to DataFusion 15.0.0

### PR DESCRIPTION
**Pull request by andygrove**
[DataFusion 15.0.0](https://github.com/apache/arrow-datafusion/issues/4468) will be released in the next few days. This PR is in preparation for upgrading.

### CODE REVIEW

File: dask_planner/src/expression.rs
Comment (andygrove) [inline dask_planner/src/expression.rs:597]: decimal scale is now a signed value

File: dask_planner/src/sql/optimizer.rs
Comment (andygrove) [inline dask_planner/src/sql/optimizer.rs:130]: Projections no longer have aliases, but use the generic `SubqueryAlias` operator instead

File: dask_sql/physical/rel/logical/subquery_alias.py
Comment (charlesbluca) [inline dask_sql/physical/rel/logical/subquery_alias.py:36]: With this DataFusion bump, we now optimize otherwise identical projections with different aliases into a single projection followed by the equivalent `SubqueryAlias`, e.g. turning

```
Projection: customer_total_return.ctr_store_sk, customer_total_return.ctr_total_return, alias=ctr2
  Projection: store_returns.sr_store_sk AS ctr_store_sk, SUM(store_returns.sr_fee) AS ctr_total_return, alias=customer_total_return
```

into

```
SubqueryAlias: ctr2
  SubqueryAlias: customer_total_return
    Projection: store_returns.sr_store_sk AS ctr_store_sk, SUM(store_returns.sr_fee) AS ctr_total_return
```

This exposed the fact that our `SubqueryAlias` plugin was a no-op that didn't actually do the required aliasing. These modifications should unblock failures stemming from this optimization.

**Review by jdye64 — APPROVED** [review]
Taking @charlesbluca comments about SubqueryAlias into account these changes LGTM

**Review by ayushdg — APPROVED** [review]
It's been a while since this pr was opened, but the changes lgtm! 
I'm holding off on merging to run some benchmarks internally but should be good to merge in a day or so

**Pull request merged**
